// -----------------------------------------------------------------------------
// Copyright 2018 Frank V. Castellucci
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";
option java_multiple_files = true;
option java_package = "hashblock.match.protobuf";

// Resource quantity
message Quantity {

    //  Quantity amount? - Fixed width integer
	bytes value = 1;

    //  Big number
	bytes valueUnit = 2;

    //  Big number
	bytes resourceUnit = 3;
}

// Quantity ratio
message Ratio {

    //  Transaction quantity of unit and resource 
    Quantity numerator = 1;

    //  Transaction quantity of unit and resource 
    Quantity denominator = 2;
}

// Economic event with plus/minus duality
message UTXQ {
    //  Keys of resource inc.
    bytes plus = 1;
    //  Keys of resource dec.
    bytes minus = 2;
    //  Transaction quantity of unit and resource 
    Quantity quantity = 3;
    //  State flag for reciprocate validity
    bool matched = 4;
}

message MTXQ {
    //  Keys of transaction inc. has to jive with IE minus
    bytes plus = 1;
    //  Keys of resource dec. has to jive with IE plus
    bytes minus = 2;
    //  Don't know how this is used
    Ratio ratio = 3;
    //  Transaction quantity of unit and resource
    Quantity quantity = 4;
    //  
    UTXQ unmatched = 5;
}

message MatchEvent {
    enum Action {
        UTXQ_ASK = 0;
        MTXQ_TELL = 1;
        UTXQ_OFFER = 2;
        MTXQ_ACCEPT = 3;
        UTXQ_COMMITMENT = 4;
        MTXQ_OBLIGATION = 5;
        UTXQ_GIVE = 6;
        MTXQ_TAKE = 7;
    }

    //  See enum above
    Action action = 1;

    //  The initiate event key
    string ukey = 2;

    //  The reciprocate event key
    string mkey = 3;

    // The content of this payload
    bytes data = 4;
}
